src/main/schema.ts

```diff
@@ -0,0 +1,120 @@
+import Ajv from 'ajv';
+import { ClientError } from './exception';
+import { ajvErrorToMessage } from './util';
+
+/**
+ * An utility class that allows validating and decoding objects using JSON Schema.
+ *
+ *
+ */
+export class Schema<T> {
+    schema: object;
+
+    protected ajv: Ajv.Ajv;
+    protected validateFn: Ajv.ValidateFunction;
+    protected defaults: SchemaDefaults<T>;
+
+    constructor(options: SchemaInit<T>) {
+        const {
+            schema,
+            defaults = {},
+        } = options;
+        this.ajv = new Ajv({
```

Comment:
Should we consider filtering additional properties? https://www.npmjs.com/package/ajv#filtering-data